### PR TITLE
feature/EVSv2-NFCENS_additions

### PR DESCRIPTION
**Pull request by SamiraArdani-NOAA**
## Pull Request Testing ##
This PR addresses the remaining items in EVS v2.0. Here, is the list of the Items:
1- Used split_by_subset to trim prepbufr files to only include needed obs before METplus.
2- Changed subprocess.Popen to subprocess.run in EVS v2.0.
3- Any use of LOOP_ORDER was removed across all components.
4- Updated config files where necessary to use METplus-6.0.0 without warnings.
5- WARNING message outside the SENDMAIL was added for missing observational data.

- [ ] Describe testing already performed for this Pull Request:</br>
I tested this PR for four rundates and set a crontab for daily runs of prep, stats and plots. Reviewers can use my prep directory located at my noscrub directory for this test:  $noscrub/samira.ardani/evs/v2.0/prep/nfcens/*.

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
2- Clone my fork in your local and checkout branch "feature/EVSv2-NFCENS_additions".
3- ln -s fix directory.
4- Run the driver scripts for prep step at: EVS/dev/drivers/scripts/prep/nfcens/*
5- Run the driver scripts for stats step at: EVS/dev/drivers/scripts/stats/nfcens/*
6- Run the driver scripts for plots step at: EVS/dev/drivers/scripts/plots/nfcens/*

- [ ] Has the code been checked to ensure that no errors occur during the execution? **[Yes ]**

- [ ] Do these updates/additions include sufficient testing updates? **[Yes]**

- [ ] Please complete this pull request review by **[08/09/2024]**.</br>

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

### CODE REVIEW

File: dev/drivers/scripts/prep/nfcens/jevs_nfcens_wave_grid2obs_prep.sh [hunk 1]
```diff
@@ -42,7 +42,7 @@ export SENDCOM=${SENDCOM:-YES}
 export SENDECF=${SENDECF:-YES}
 export SENDDBN=${SENDDBN:-NO}
 export KEEPDATA=${KEEPDATA:-YES}
-export SENDMAIL=${SENDMAIL:-YES}
+export SENDMAIL=${SENDMAIL:-NO}
```
Comment (AliciaBentley-NOAA) [inline dev/drivers/scripts/prep/nfcens/jevs_nfcens_wave_grid2obs_prep.sh]: @SamiraArdani-NOAA Can you please change SENDMAIL to YES in your dev driver (unless there are usually a huge number of missing file emails)? You can set SENDMAIL to NO in your J-job, if it isn't already! 

CC @malloryprow @ShelleyMelchior-NOAA

**Comment by malloryprow** [comment]
I'm going to be working on PR Samira! I have some administrative work tasks to do after lunch so I'll start on this tomorrow. In the meantime if you could merge the most recent changes in develop into your branch that be great.

**Comment by malloryprow** [comment]
I'm having trouble with the training I was going to take today, so getting started on this PR!

**Comment by malloryprow** [comment]
I have run jevs_nfcens_wave_grid2obs_prep.sh with  INITDATE=20240729 and  NEXTDATE=20240730. The job is complete.

**Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/EVS/dev/drivers/scripts/prep/nfcens/jevs_nfcens_wave_grid2obs_prep.o146326870
**COMOUT:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/evs/v2.0/prep/nfcens/wave
**DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01

I did receive 12 missing data emails. Here is an example of one of them:

> WARNING: No FNMOC forecast was available for 20240729
> WARNING: Missing file is /lfs/h1/ops/prod/com/fnmoc_wave/v1.1/fnmoc_wave.20240729/wave_2024072900f114
> Job ID: jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01

Is this expected?

**Comment by malloryprow** [comment]
Additionally, I noticed the pb2nc run is not using the output from split_by_subset. We want the input of pb2nc to run on the output produced from running split_by_subset.

> DEBUG 1: Start pb2nc by mallory.row(32288) at 2024-07-31 16:51:49Z  cmd: /apps/ops/para/libs/intel/19.1.3.304/met/12.0.0-beta4/bin/pb2nc -v 2 /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01/gdas.2024072912.prepbufr /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01/SFCSHP/gdas.SFCSHP.2024072912.nc /apps/ops/para/libs/intel/19.1.3.304/metplus/6.0.0-beta4/parm/met_config/PB2NCConfig_wrapped

I see that split_by_subset is being run.

> split_by_subset /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01/gdas.2024072912.prepbufr

and it looks like that output is in /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01/gribs.

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow : Those missing data emails for fnmoc_wave are expected because of the two reasons: 1- no specific pattern in the populated data. the fnmoc_wave forecast data is populated every 3 hours up to f024, then up to 6 hours up to f072 and then every 12 hours up to f144. 2- The missing is very normal for the fnmoc_wave. Sometimes no data is available at all and sometimes one cycle is missing.

**Comment by SamiraArdani-NOAA** [comment]
> I have run jevs_nfcens_wave_grid2obs_prep.sh with INITDATE=20240729 and INITDATE=20240729. The job is complete.
> 
> **Log File:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/EVS/dev/drivers/scripts/prep/nfcens/jevs_nfcens_wave_grid2obs_prep.o146326870 **COMOUT:** /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/evs/v2.0/prep/nfcens/wave **DATA:** /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01
> 
> I did receive 12 missing data emails. Here is an example of one of them:
> 
> > WARNING: No FNMOC forecast was available for 20240729
> > WARNING: Missing file is /lfs/h1/ops/prod/com/fnmoc_wave/v1.1/fnmoc_wave.20240729/wave_2024072900f114
> > Job ID: jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01
> 
> Is this expected?

@malloryprow : I checked the prep .o files. I did not observe ERRORs. I see some WARNINGs due to the missing fnmoc_wave data which are expected. the output results look identical to emc.vpppg in terms of size. The name of those two .nc files are different as expected due to the changes for this PR.

**Comment by SamiraArdani-NOAA** [comment]
> Additionally, I noticed the pb2nc run is not using the output from split_by_subset. We want the input of pb2nc to run on the output produced from running split_by_subset.
> 
> > DEBUG 1: Start pb2nc by mallory.row(32288) at 2024-07-31 16:51:49Z  cmd: /apps/ops/para/libs/intel/19.1.3.304/met/12.0.0-beta4/bin/pb2nc -v 2 /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01/gdas.2024072912.prepbufr /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01/SFCSHP/gdas.SFCSHP.2024072912.nc /apps/ops/para/libs/intel/19.1.3.304/metplus/6.0.0-beta4/parm/met_config/PB2NCConfig_wrapped
> 
> I see that split_by_subset is being run.
> 
> > split_by_subset /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01/gdas.2024072912.prepbufr
> 
> and it looks like that output is in /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146326870.cbqs01/gribs.

@malloryprow : I am looking into this now.

**Comment by malloryprow** [comment]
New changes have been pulled in and I will rerun prep with these changes.

**Comment by malloryprow** [comment]
I have run jevs_nfcens_wave_grid2obs_prep.sh with INITDATE=20240730 and NEXTDATE=20240731. The job is complete.

**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/EVS/dev/drivers/scripts/prep/nfcens/jevs_nfcens_wave_grid2obs_prep.o146389779
**COMOUT**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/evs/v2.0/prep/nfcens/wave
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146389779.cbqs01

I'm seeing 

> 08/01 12:07:27.083 metplus.16efffa2 INFO: COMMAND: /apps/ops/para/libs/intel/19.1.3.304/met/12.0.0-beta4/bin/pb2nc -v 2 /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146389779.cbqs01/gribs/SFCSHP /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.146389779.cbqs01/SFCSHP/gdas.SFCSHP.2024073012.nc /apps/ops/para/libs/intel/19.1.3.304/metplus/6.0.0-beta4/parm/met_config/PB2NCConfig_wrapped

👍

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow : I checked the .o files and the outputs. They look good. Also, the the usage of "split_by_subset" was corrected.

**Comment by malloryprow** [comment]
Great news! I left a comment on some additions for the prep ex-script, but I feel it is okay to move on to the stats portion. Is there somewhere I can set COMIN to?

**Comment by SamiraArdani-NOAA** [comment]
> Great news! I left a comment on some additions for the prep ex-script, but I feel it is okay to move on to the stats portion. Is there somewhere I can set COMIN to?

@malloryprow : Yes, I have 10 days of data for this PR. You can set your COMIN to: /lfs/h2/emc/vpppg/noscrub/samira.ardani/evs/v2.0/prep/nfcens/

**Comment by malloryprow** [comment]
I have run jevs_nfcens_wave_grid2obs_stats.sh with VDATE=20240730 and NEXTDATE=20240731. The job is complete.

**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/EVS/dev/drivers/scripts/stats/nfcens/jevs_nfcens_wave_grid2obs_stats.o146391389
**COMOUT**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/evs/v2.0/prep/nfcens/wave
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_stats.146391389.cbqs01

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow : No ERROR was found in .o file. The stats outputs were checked. I see the differences in the size of stats from fnmoc_wave when I compare with those in emc.vpppg but I think this is expected. Since we use split_by_subset, only the data from SFCSHP was analyzed and was written in the stats whereas in emc.vpppg we have both SFCSHP and ADPSFC.

**Comment by malloryprow** [comment]
Thanks Samira! I'm dealing with an ops failure currently. Should the NFCENS only be using SFCSHP observations? (I feel like yes since it is a wave model but want to be sure) @AliciaBentley-NOAA

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow : Yes, we only use SFCSHP. I am not in rush for this PR, so, take your time for the ops failure. I completely understand.

**Comment by malloryprow** [comment]
Great to hear the differences are what is expected. I'll move on to plots. Do you have somewhere you would like me to point COMIN to?

**Comment by SamiraArdani-NOAA** [comment]
> Great to hear the differences are what is expected. I'll move on to plots. Do you have somewhere you would like me to point COMIN to?

@malloryprow, Yes. You can use /lfs/h2/emc/vpppg/noscrub/samira.ardani/evs/v2.0/stats/nfcens/*. However, the outputs at this path do not reflect the changes that I made yesterday within this PR for split_by_subset. The good thing is that this change only affected the fnmoc_wave stats.

**Comment by malloryprow** [comment]
I ran jevs_nfcens_wave_grid2obs_plots.sh with VDATE=20240730 and PDYm90=20240502.

**Log File**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/EVS/dev/drivers/scripts/plots/nfcens/jevs_nfcens_wave_grid2obs_plots.o146423290
**DATA**: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_plots.146423290.cbqs01
**COMOUT**: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/evs/v2.0

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow, I am checking the outputs now.

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow, I untarred the plots and checked the number of plots with those in emc.vpppg. The number of plots looks the same as emc.vpppg.

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow, Also, I did not see any ERROR message in the .o file.

**Comment by malloryprow** [comment]
Great!

I added some comments about file checking in the prep script. Before I approve this I would like to see that added and then re-test prep.

Additionally, I think prep should be enhanced for FNMOC wave data. It sounds like there is data it is looking for that is never expected to be there (data available every 3 hours up to f24, every 6 hours to f72, every 12 hours up to f144). NCO isn't going to like WARNING message for data that will never be there. I would suggest a `do` loop instead of a `while` loop. This is something that can be a different PR.

**Comment by SamiraArdani-NOAA** [comment]
> Great!
> 
> I added some comments about file checking in the prep script. Before I approve this I would like to see that added and then re-test prep.
> 
> Additionally, I think prep should be enhanced for FNMOC wave data. It sounds like there is data it is looking for that is never expected to be there (data available every 3 hours up to f24, every 6 hours to f72, every 12 hours up to f144). NCO isn't going to like WARNING message for data that will never be there. I would suggest a `do` loop instead of a `while` loop. This is something that can be a different PR.

@malloryprow : You are right. I was thinking about that too. I guess this should be the topic of next PRs. Right?

**Comment by malloryprow** [comment]
Yes, we can add that to the Fixes and Additions document. @AliciaBentley-NOAA for awareness!

**Comment by AliciaBentley-NOAA** [comment]
@malloryprow @SamiraArdani-NOAA I added this to the [Fixes and Additions in EVS v2.0](https://docs.google.com/document/d/1vOGnCwCM3COUhQlTqHyBZrYSCPx9b6AYnxLc5NtVGEk/edit#heading=h.q8i76uwez5nr) document! Please let me know if I should add anything else added!

**Review by AliciaBentley-NOAA — APPROVED** [review]
I have reviewed the changes made in this PR and noted what has been completed on the Fixes and Additions document. If you can confirm that there are no WARNINGS associated with using MET-12.0.0/METplus-6.0.0, I can check that off the Fixes and Additions document as well! @malloryprow

File: scripts/prep/nfcens/exevs_nfcens_wave_grid2obs_prep.sh
Comment (malloryprow) [inline scripts/prep/nfcens/exevs_nfcens_wave_grid2obs_prep.sh:1]: There should be some file checking here to help avoid errors and potential future failures in ops to keep NCO happy.

Before split_by_subset check for `${DATA}/gdas.${INITDATE}${HH}.prepbufr`
Before run_metplus.py, check for `${DATA}/gribs/SFCSHP`
Before the copy to COMOUT, check for `$DATA/SFCSHP/gdas.SFCSHP.${INITDATE}${HH}.nc`

**Comment by malloryprow** [comment]
> I have reviewed the changes made in this PR and noted what has been completed on the Fixes and Additions document. If you can confirm that there are no WARNINGS associated with using MET-12.0.0/METplus-6.0.0, I can check that off the Fixes and Additions document as well! @malloryprow

Yup! Only WARNING messages I am seeing are due to FNMOC wave missing data.

**Comment by AliciaBentley-NOAA** [comment]
@malloryprow @SamiraArdani-NOAA Perfection. I'll check that off in the Fixes and Additions document too!

**Comment by malloryprow** [comment]
@SamiraArdani-NOAA Do you have an update for how adding the file checking is going?

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow : I thought I had to fix it in the next PR. Do you want me to fix it in this PR?

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow : I can do it in this PR though if it works for you.

**Comment by malloryprow** [comment]
Yes, please fix in this PR. 

The changes for the forecast hour for FNMOC wave we can address in a separate PR.

**Comment by malloryprow** [comment]
I have run jevs_nfcens_wave_grid2obs_prep.sh with INITDATE=20240805 and NEXTDATE=20240806. The job is complete.

Log File: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/EVS/dev/drivers/scripts/prep/nfcens/jevs_nfcens_wave_grid2obs_prep.o147278916
COMOUT: /lfs/h2/emc/vpppg/noscrub/mallory.row/verification/EVS_PRs/pr504/evs/v2.0/prep/nfcens/wave
DATA: /lfs/h2/emc/stmp/mallory.row/evs_test/prod/tmp/jevs_nfcens_wave_grid2obs_prep.147278916.cbqs01

**Comment by SamiraArdani-NOAA** [comment]
@malloryprow 
- No ERROR was found in .o file.
- The output results from models were identical with emc.vpppg. The final output results from pb2nc had less sizes compare to emc.vpppg (I guess it is expected due to the use of split_by_subset).

**Review by malloryprow — APPROVED** [review]

**Pull request merged**